### ISSUE TITLE
DASH crashed while writing file failed.

### ISSUE THREAD

**Issue by winlinvip** [ossrs/srs#1545]
**环境(Environment)**

1. SRS版本(Version): `3.0.85`
1. SRS的日志如下(Log):
```

[2019-12-27 03:05:05.214][Warn][9287][752][104] client disconnect peer. ret=1007
[2019-12-27 03:05:08.066][Warn][9287][746][14] dash: ignore video error code=1045 : Consume video failed : reap current : Flush encoder failed : write sample : write to file ./objs/nginx/html/live/livestream/video-525805301.m4s.tmp failed
thread [9287][746]: on_video() [src/app/srs_app_dash.cpp:558][errno=14]
thread [9287][746]: on_video() [src/app/srs_app_dash.cpp:406][errno=14]
thread [9287][746]: reap() [src/app/srs_app_dash.cpp:149][errno=14]
thread [9287][746]: flush() [src/kernel/srs_kernel_mp4.cpp:6335][errno=14]
thread [9287][746]: write() [src/kernel/srs_kernel_file.cpp:137][errno=14]
*** Error in `/tmp/srs/trunk/./objs/srs': free(): invalid pointer: 0x0000000000be57b5 ***

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff73c70b8 in ?? () from /lib64/libgcc_s.so.1
Missing separate debuginfos, use: debuginfo-install glibc-2.17-292.el7.x86_64 libgcc-4.8.5-39.el7.x86_64 libstdc++-4.8.5-39.el7.x86_64
(gdb) bt
#0  0x00007ffff73c70b8 in ?? () from /lib64/libgcc_s.so.1
#1  0x00007ffff73c7fb9 in _Unwind_Backtrace () from /lib64/libgcc_s.so.1
#2  0x00007ffff70fea56 in backtrace () from /lib64/libc.so.6
#3  0x00007ffff7062ea4 in __libc_message () from /lib64/libc.so.6
#4  0x00007ffff706b679 in _int_free () from /lib64/libc.so.6
#5  0x0000000000438ea5 in SrsMp4Sample::~SrsMp4Sample (this=0xbe7e30, __in_chrg=<optimized out>) at src/kernel/srs_kernel_mp4.cpp:4593
#6  0x0000000000438ee0 in SrsMp4Sample::~SrsMp4Sample (this=0xbe7e30, __in_chrg=<optimized out>) at src/kernel/srs_kernel_mp4.cpp:4594
#7  0x0000000000438ff6 in SrsMp4SampleManager::~SrsMp4SampleManager (this=0xabe8e0, __in_chrg=<optimized out>) at src/kernel/srs_kernel_mp4.cpp:4615
#8  0x00000000004390aa in SrsMp4SampleManager::~SrsMp4SampleManager (this=0xabe8e0, __in_chrg=<optimized out>) at src/kernel/srs_kernel_mp4.cpp:4618
#9  0x0000000000440d6c in SrsMp4M2tsSegmentEncoder::~SrsMp4M2tsSegmentEncoder (this=0xbd4d40, __in_chrg=<optimized out>)
    at src/kernel/srs_kernel_mp4.cpp:6169
#10 0x0000000000440dec in SrsMp4M2tsSegmentEncoder::~SrsMp4M2tsSegmentEncoder (this=0xbd4d40, __in_chrg=<optimized out>)
    at src/kernel/srs_kernel_mp4.cpp:6171
#11 0x000000000058b2bd in SrsFragmentedMp4::~SrsFragmentedMp4 (this=0xba0d20, __in_chrg=<optimized out>) at src/app/srs_app_dash.cpp:80
#12 0x000000000058b36a in SrsFragmentedMp4::~SrsFragmentedMp4 (this=0xba0d20, __in_chrg=<optimized out>) at src/app/srs_app_dash.cpp:82
#13 0x000000000058da4c in SrsDashController::on_unpublish (this=0xa5f670) at src/app/srs_app_dash.cpp:359
#14 0x000000000058eade in SrsDash::on_unpublish (this=0xa5fcc0) at src/app/srs_app_dash.cpp:573
#15 0x00000000004cd2c6 in SrsOriginHub::on_video (this=0xa5f2c0, shared_video=0x7ffff7ef1b20, is_sequence_header=false) at src/app/srs_app_source.cpp:1073
#16 0x00000000004d2980 in SrsSource::on_video_imp (this=0xa5e320, msg=0x7ffff7ef1b20) at src/app/srs_app_source.cpp:2239
#17 0x00000000004d25fa in SrsSource::on_video (this=0xa5e320, shared_video=0xba8f70) at src/app/srs_app_source.cpp:2194
#18 0x00000000004c68e8 in SrsRtmpConn::process_publish_message (this=0xa54650, source=0xa5e320, msg=0xba8f70) at src/app/srs_app_rtmp_conn.cpp:1020
#19 0x00000000004c6733 in SrsRtmpConn::handle_publish_message (this=0xa54650, source=0xa5e320, msg=0xba8f70) at src/app/srs_app_rtmp_conn.cpp:992
#20 0x00000000005637e2 in SrsPublishRecvThread::consume (this=0x7ffff7f4b810, msg=0xba8f70) at src/app/srs_app_recv_thread.cpp:389
#21 0x000000000056251f in SrsRecvThread::do_cycle (this=0x7ffff7f4b818) at src/app/srs_app_recv_thread.cpp:146
#22 0x000000000056237f in SrsRecvThread::cycle (this=0x7ffff7f4b818) at src/app/srs_app_recv_thread.cpp:115
#23 0x00000000004f390e in SrsSTCoroutine::cycle (this=0xa5f190) at src/app/srs_app_st.cpp:198
#24 0x00000000004f3983 in SrsSTCoroutine::pfn (arg=0xa5f190) at src/app/srs_app_st.cpp:213
#25 0x000000000059a8a6 in _st_thread_main () at sched.c:337
#26 0x000000000059b016 in st_thread_create (start=
    0x531c9b <std::_Rb_tree<int, std::pair<int const, SrsStageInfo*>, std::_Select1st<std::pair<int const, SrsStageInfo*> >, std::less<int>, std::allocator<std::pair<int const, SrsStageInfo*> > >::end()+35>, arg=0x7ffff7f4b590, joinable=32767, stk_size=-134957696) at sched.c:616
#27 0x00000003ffffffff in ?? ()
#28 0x00007ffff7f7ee98 in ?? ()
#29 0x0000000000531ae1 in std::map<int, SrsStageInfo*, std::less<int>, std::allocator<std::pair<int const, SrsStageInfo*> > >::operator[] (
    this=<error reading variable: Cannot access memory at address 0x4ffffff9d>, __k=<error reading variable: Cannot access memory at address 0x4ffffff95>)
    at /usr/include/c++/4.8.2/bits/stl_map.h:471
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

**Comment by winlinvip** [comment]
Check it.

**Comment by winlinvip** [comment]
Fixed in 5.0.96, see https://github.com/ossrs/srs/pull/3240
